File: 210/level3/가장먼노드.java
// 20220819 :: 20분 컷인줄 알았는데... Queue의 자료형을 LinkedList이 아닌 PriorityQueue를 아무 생각 없이 쓴게 삽질 원인ㅜㅜ

package level3;

import java.util.*;

class Main {
    public static void main(String[] args) throws Exception {
        int n = 11;
        int[][] edge = { { 1, 2 }, { 1, 3 }, { 2, 4 }, { 2, 5 }, { 3, 5 }, { 3, 6 }, { 4, 8 }, { 4, 9 }, { 5, 9 },
                { 5, 10 }, { 6, 10 }, { 6, 11 } };

        Solution sol = new Solution();
        System.out.println("result : " + sol.solution(n, edge));
    }
}

class Solution {
    public int solution(int n, int[][] edge) {
        int maxDis = -1, answer = 0, edgeCnt = edge.length;
        int[] distance = new int[n+1];
        boolean[] visited = new boolean[n+1];
        ArrayList<Integer>[] edges = new ArrayList[n + 1];
        Queue<Integer> q = new LinkedList<>();

        for (int i = 1; i <= n; i++) {
            distance[i] = 50000;
            edges[i] = new ArrayList<Integer>();
        }

        for (int i = 0; i < edgeCnt; i++) {
            edges[edge[i][0]].add(edge[i][1]);
            edges[edge[i][1]].add(edge[i][0]);
        }

        visited[1] = true;
        distance[1] = 0;
        q.add(1);
        while (!q.isEmpty()) {
            int currNum = q.poll();
            visited[currNum] = true;
            for (int i = 0; i < edges[currNum].size(); i++) {
                if (!visited[edges[currNum].get(i)]) {
                    distance[edges[currNum].get(i)] = Math.min(distance[edges[currNum].get(i)], distance[currNum] + 1);
                    maxDis = Math.max(maxDis, distance[edges[currNum].get(i)]);
                    q.add(edges[currNum].get(i));
                }
            }
        }

        for (int i = 1; i <= n; i++) {
            if (distance[i] == maxDis) answer++;
        }

        return answer;
    }
}

File: JunJangE/level3/가장 먼 노드.kt
package level3

import java.util.*

fun main() {
    println(Solution().solution(6, arrayOf(intArrayOf(3, 6), intArrayOf(4, 3), intArrayOf(3, 2), intArrayOf(1, 3), intArrayOf(1, 2), intArrayOf(2, 4), intArrayOf(5, 2))))
}


class Solution {

    fun solution(n: Int, edge: Array<IntArray>): Int {
        var answer : Int = 0
        val graph : Array<ArrayList<Int>> = Array(n + 1) { ArrayList<Int>() }
        val visited : IntArray = IntArray(n+1){0}
        val queue : Queue<Int> = LinkedList<Int>()

        edge.forEach {
            graph[it[0]].add(it[1])
            graph[it[1]].add(it[0])

        }

        queue.add(1)
        visited[1] = 1


        while (queue.isNotEmpty()){
            val x = queue.poll()

            for (i in graph[x]){
                if (visited[i] == 0){
                    visited[i] = visited[x] + 1
                    queue.add(i)
                }
            }

        }

        //가장 먼 노드의 depth
        visited.sortDescending()

        val depthMax = visited[0]
        answer = visited.count { it == depthMax }
        return answer
    }
}
